diff --git a/ext/session/mod_files.c b/ext/session/mod_files.c
--- a/ext/session/mod_files.c
+++ b/ext/session/mod_files.c
@@ -199,7 +199,11 @@ static void ps_files_open(ps_files *data, const char *key)
                accessed by backend with root permissions to execute some system tasks.
 
                */
-            if (zend_fstat(data->fd, &sbuf) || (sbuf.st_uid != 0 && sbuf.st_uid != getuid() && sbuf.st_uid != geteuid() && getuid() != 0)) {
+#ifdef __wasi__
+            if (zend_fstat(data->fd, &sbuf)) {
+#else
+            if (zend_fstat(data->fd, &sbuf) || (sbuf.st_uid != 0 && sbuf.st_uid != getuid() && sbuf.st_uid != geteuid() && getuid() != 0)) {
+#endif
                 close(data->fd);
                 data->fd = -1;
                 php_error_docref(NULL, E_WARNING, "Session data file is not created by your uid");
